### PR TITLE
allow using apex with any PT version

### LINKED ISSUE #2788: Enable option to use Apex when PyTorch 1.6 is installed
## 🚀 Feature

Currently, if PyTorch 1.6 is installed, PyTorch Lightning will use native AMP by default. It would be useful to have an option to force PL to use Apex if needed, which is not currently possible

### Motivation

Native AMP is great but under certain conditions may perform differently to Apex. For debugging purposes and backwards compatibility, having the option to still use Apex might be useful.

In my particular problem, I needed to roll back to PT 1.5 since I was seeing different behaviour between the two. Having the option to switch between the two AMP methods would make it easier to diagnose my issue.

### Pitch

Have a flag to force the Trainer to use Apex over native AMP

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

add extra argument which would switch between using Apex and native AMP in pt>=1.6

Fixes #2788

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/memory.py [hunk 1]
```diff
@@ -207,8 +207,8 @@ def _forward_example_input(self) -> None:
         input_ = model.example_input_array
         input_ = model.transfer_batch_to_device(input_, model.device)
 
-        if trainer is not None and trainer.use_amp and not trainer.use_tpu:
-            if NATIVE_AMP_AVALAIBLE:
+        if trainer is not None and trainer.amp_type and not trainer.use_tpu:
+            if trainer.amp_type == AMPType.NATIVE:
```
Comment (awaelchli) [inline pytorch_lightning/core/memory.py:211]:
```suggestion
        if trainer is not None and trainer.amp_type == AMPType.NATIVE and not trainer.use_tpu:
```

File: pytorch_lightning/trainer/__init__.py [hunk 2]
```diff
@@ -864,6 +864,18 @@ def on_train_end(self, trainer, pl_module):
 
     trainer = Trainer(sync_batchnorm=True)
 
+amp_type
+^^^^^^^^
+
+Define of preferable mixed precion - NVIDIA Apex ("apex") or PyTorch build-in ("native") supported from v1.6.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/__init__.py:870]:
```suggestion
Define a preferable mixed precision, either NVIDIA Apex ("apex") or PyTorch built-in ("native") AMP which is supported from v1.6.
```

File: pytorch_lightning/trainer/__init__.py [hunk 3]
```diff
@@ -864,6 +864,18 @@ def on_train_end(self, trainer, pl_module):
...
+
+.. testcode::
+
+    # using NVIDIA Apex
+    trainer = Trainer(amp_type='apex')
+    # using PyTorch build-in
+    trainer = Trainer(amp_type='native')
```
Comment (awaelchli) [inline pytorch_lightning/trainer/__init__.py:877]:
```suggestion
    # using NVIDIA Apex
    trainer = Trainer(amp_type='apex')
    
    # using PyTorch built-in AMP
    trainer = Trainer(amp_type='native')
```

File: pytorch_lightning/trainer/auto_mix_precision.py [hunk 4]
```diff
@@ -10,26 +10,39 @@ class TrainerAMPMixin(ABC):
...
+                log.info('Using APEX 16bit precision.')
+                self.amp_type = AMPType.APEX
+        if not self.amp_type:
             raise ModuleNotFoundError(
-                "You set `use_amp=True` but do not have apex installed."
-                " Install apex first using this guide: https://github.com/NVIDIA/apex#linux"
-                " and rerun with `use_amp=True`."
-                " This run will NOT use 16 bit precision."
+                f'You have asked for AMP support {amp_type} there is no support on your side yet.'
+                f' Consider install Trorch>=1.6 or NVIDIA Apex.'
```
Comment (awaelchli) [inline pytorch_lightning/trainer/auto_mix_precision.py:39]:
```suggestion
                f'You have asked for AMP support {amp_type}, but there is no support on your side yet.'
                f' Consider installing torch >= 1.6 or NVIDIA Apex.'
```

File: pytorch_lightning/trainer/auto_mix_precision.py [hunk 5]
```diff
@@ -10,26 +10,39 @@ class TrainerAMPMixin(ABC):
...
             return
-
-        if self.use_amp and not APEX_AVAILABLE:  # pragma: no-cover
+        amp_type = amp_type.lower()
+        assert amp_type in ('native', 'apex'), f'Unsupported amp source {amp_type}'
+        if amp_type == 'native':
+            if not NATIVE_AMP_AVALAIBLE:
+                rank_zero_warn('You have asked for native AMP but your PyTorch version does not support it.'
+                               ' Consider upgrading `pip install torch>=1.6`.'
+                               ' Lets try to use NVIDIA Apex for this session.')
```
Comment (awaelchli) [inline pytorch_lightning/trainer/auto_mix_precision.py:24]:
```suggestion
                rank_zero_warn('You have asked for native AMP but your PyTorch version does not support it.'
                               ' Consider upgrading with `pip install torch>=1.6`.'
                               ' We will attempt to use NVIDIA Apex for this session.')
```

File: pytorch_lightning/trainer/auto_mix_precision.py [hunk 6]
```diff
@@ -10,26 +10,39 @@ class TrainerAMPMixin(ABC):
...
-            log.info('Using native 16bit precision.')
+    def _setup_amp_type(self, amp_type: str):
+        self.amp_type = None
+        if self.precision != 16:
+            # no AMP requested, so we can leave now
             return
-
-        if self.use_amp and not APEX_AVAILABLE:  # pragma: no-cover
+        amp_type = amp_type.lower()
+        assert amp_type in ('native', 'apex'), f'Unsupported amp source {amp_type}'
```
Comment (awaelchli) [inline pytorch_lightning/trainer/auto_mix_precision.py:19]:
```suggestion
        assert amp_type in ('native', 'apex'), f'Unsupported amp type {amp_type}'
```

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -199,6 +197,7 @@ def __init__(
         terminate_on_nan: bool = False,
         auto_scale_batch_size: Union[str, bool] = False,
         prepare_data_per_node: bool = True,
+        amp_type: str = 'native',
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:200]: was this the default before? let's make sure it's correct in the docs.

File: pytorch_lightning/utilities/__init__.py [hunk 8]
```diff
@@ -19,3 +20,8 @@
 FLOAT16_EPSILON = numpy.finfo(numpy.float16).eps
 FLOAT32_EPSILON = numpy.finfo(numpy.float32).eps
 FLOAT64_EPSILON = numpy.finfo(numpy.float64).eps
+
+
+class AMPType(Enum):
+    APEX = 'apex'
+    NATIVE = 'native'
```
Comment (awaelchli) [inline pytorch_lightning/utilities/__init__.py:27]: I didn't find the place where you convert the string trainer arg to this type, where is it?

**Comment by awaelchli** [comment]
@Borda I like this PR a lot!! Nice simplifications!

File: pytorch_lightning/accelerators/ddp_backend.py [hunk 9]
```diff
@@ -202,10 +200,8 @@ def ddp_train(self, process_idx, mp_queue, model, is_master=False, proc_offset=0
         # set model properties before going into wrapper
         self.trainer.copy_trainer_model_properties(model)
 
-        # AMP
-        # run through amp wrapper before going to distributed DP
-        # TODO: remove with dropping NVIDIA AMP support
-        if self.trainer.use_amp and not NATIVE_AMP_AVALAIBLE:
+        # AMP - run through amp wrapper before going to distributed DP
+        if self.trainer.amp_type == AMPType.APEX:
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/accelerators/ddp_backend.py:204]: no, the point of this PR is to keep Apex support side by side with native amp

[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/accelerators/ddp_backend.py:204]: no, we keep apex and native side by side

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:200]: before it uses native if available and if not, try apex... not you can choose on your own

File: pytorch_lightning/accelerators/ddp_backend.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/accelerators/ddp_backend.py:204]: ok, then you need to remove the todo in some other places, I think ddp2_backend has it still

File: pytorch_lightning/trainer/training_loop.py [hunk 10]
```diff
@@ -954,18 +956,18 @@ def optimizer_closure(self, split_batch, batch_idx, opt_idx, optimizer, hiddens)
         with self.profiler.profile('model_backward'):
             # scale loss for 16 bit
             if self.precision == 16 and not self.on_tpu:
-                closure_loss = model_ref.amp_scale_loss(closure_loss, optimizer, opt_idx)
+                closure_loss = model_ref.amp_scale_loss(closure_loss, optimizer, opt_idx, amp_type=self.amp_type)
 
                 # enter amp context
-                if not NATIVE_AMP_AVALAIBLE:
+                if amp_type == AMPType.APEX:
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:962]:
```suggestion
                if self.amp_type == AMPType.APEX:
```

File: pytorch_lightning/utilities/__init__.py (hunk 8 above)
Comment (Borda) [reply pytorch_lightning/utilities/__init__.py:27]: It is in `init_amp` in AMP mixin...

**Comment by Borda** [comment]
Todo, remove all amp todoss... https://github.com/PyTorchLightning/pytorch-lightning/pull/2865#discussion_r467373674

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (justusschock) [inline pytorch_lightning/trainer/trainer.py:200]:
```suggestion
        amp_type: Union[str, AMPType] = 'native',
```

File: pytorch_lightning/trainer/training_io.py [hunk 11]
```diff
@@ -157,8 +155,9 @@ class TrainerIOMixin(ABC):
     on_tpu: bool
     num_training_batches: int
     accumulate_grad_batches: int
-    use_amp: bool
     scaler: ...
+    use_tpu: bool
+    amp_type: AMPType
```
Comment (justusschock) [inline pytorch_lightning/trainer/training_io.py:160]:
```suggestion
    amp_type: Union[str, AMPType]
```

File: pytorch_lightning/trainer/training_loop.py [hunk 12]
```diff
@@ -255,6 +253,8 @@ class TrainerTrainLoopMixin(ABC):
     terminate_on_nan: bool
     tpu_id: int
     interactive_ddp_procs: ...
+    amp_type: AMPType
```
Comment (justusschock) [inline pytorch_lightning/trainer/training_loop.py:256]:
```suggestion
    amp_type: Union[str, AMPType]
```

File: pytorch_lightning/trainer/training_tricks.py [hunk 13]
```diff
@@ -48,6 +46,7 @@ class TrainerTrainingTricksMixin(ABC):
     default_root_dir: str
     progress_bar_callback: ...
     on_gpu: bool
+    amp_type: AMPType
```
Comment (justusschock) [inline pytorch_lightning/trainer/training_tricks.py:49]:
```suggestion
    amp_type: Union[str, AMPType]
```

File: pytorch_lightning/trainer/training_io.py (hunk 11 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_io.py:160]: no, here just the type...

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:200]: lets use here only string...

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:200]: got it!!

**Comment by awaelchli** [comment]
@Borda I linked #2788, ok?

**Comment by Borda** [comment]
> btw, you can directly parse the input args to a type like this
> 
> ```python
> self.amp_type = AMPType[amp_type.upper()]
> ```
> 
> anyway, I like this PR :)

Cool! that is nice, but in this case, we also need to verify if it is available...

**Pull request merged**